module/VuFind/src/VuFind/RateLimiter/RateLimiterManager.php

```diff
@@ -0,0 +1,299 @@
+                $headers = $event->getResponse()->getHeaders();
+                $headers->addHeaders(
+                    [
+                        'X-RateLimit-Remaining' => $result['requestsRemaining'],
+                        'X-RateLimit-Retry-After' => $result['retryAfter'],
+                        'X-RateLimit-Limit' => $result['requestLimit'],
+                    ]
+                );
+            }
+            if ($limit->isAccepted()) {
+                return $result;
+            }
+            $logMsg = "$this->eventDesc: $this->clientId policy '$policyId' exceeded";
+            if ('report_only' === $this->isEnabled() || ($this->config['Policies'][$policyId]['reportOnly'] ?? false)) {
+                $this->logWarning("$logMsg (not enforced)");
+                return $result;
+            }
+            $this->logWarning("$logMsg (enforced)");
+            $result['allow'] = false;
+            $result['message'] = $this->getTooManyRequestsResponseMessage($event, $result);
+            return $result;
+        } catch (\Exception $e) {
+            $this->logError((string)$e);
+        }
+        // Allow access on failure:
+        return $result;
+    }
+
+    /**
+     * Try to find a policy that matches an event
+     *
+     * @param MvcEvent $event Event
+     *
+     * @return ?string policy id or null if no match
+     */
+    protected function getPolicyIdForEvent(MvcEvent $event): ?string
+    {
+        foreach ($this->config['Policies'] ?? [] as $name => $settings) {
+            if (null !== ($loggedIn = $settings['loggedIn'] ?? null)) {
+                if ($loggedIn !== ($this->userId ? true : false)) {
+                    continue;
+                }
+            }
+            if ($ipRanges = $settings['ipRanges'] ?? null) {
+                if (!$this->ipUtils->isInRange($this->clientIp, (array)$ipRanges)) {
+                    continue;
+                }
+            }
+
+            if (!($filters = $settings['filters'] ?? null)) {
+                return $name;
+            }
+            foreach ($filters as $filter) {
+                if ($this->eventMatchesFilter($event, $filter)) {
+                    return $name;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Check if an event matches a filter
+     *
+     * @param MvcEvent $event  Event
+     * @param array    $filter Filter from configuration
+     *
+     * @return bool
+     */
+    protected function eventMatchesFilter(MvcEvent $event, array $filter): bool
+    {
+        $routeMatch = $event->getRouteMatch();
+        foreach ($filter as $param => $value) {
+            if ('name' === $param) {
+                if ($routeMatch?->getMatchedRouteName() !== $value) {
+                    return false;
+                }
+            } elseif (in_array($param, ['params', 'query', 'post'])) {
+                $req = $event->getRequest();
+                if ('params' === $param) {
```

Comment:
This might be a good use case for a [match expression](https://www.php.net/manual/en/control-structures.match.php).

Reply:
True, changed!